Source/WebCore/Modules/webaudio/BaseAudioContext.cpp

```diff
@@ -298,16 +299,18 @@ void BaseAudioContext::decodeAudioData(Ref<ArrayBuffer>&& audioData, RefPtr<Audi
     if (!m_audioDecoder)
         m_audioDecoder = makeUnique<AsyncAudioDecoder>();
 
-    m_audioDecoder->decodeAsync(WTFMove(audioData), sampleRate(), [this, activity = makePendingActivity(*this), successCallback = WTFMove(successCallback), errorCallback = WTFMove(errorCallback), promise = WTFMove(promise)](ExceptionOr<Ref<AudioBuffer>>&& result) mutable {
+    // We don't move audioData as we want to delete it on the main thread upon completion of the DecodingTaskPromise.
+    auto p = m_audioDecoder->decodeAsync(Ref { audioData }, sampleRate());
+    p->then(RunLoop::main(), __func__, [this, activity = makePendingActivity(*this), successCallback = WTFMove(successCallback), errorCallback = WTFMove(errorCallback), promise = WTFMove(promise), audioData = WTFMove(audioData)] (DecodingTaskPromise::Result&& result) mutable {
```

Comment:
Why capturing audioData in the then callback?
Can we just keep passing/moving it to decodeAsync?

Reply:
>Why capturing audioData in the then callback?
I explained the why in the comment above.
audioData is a JS array buffer, it's main thread only.

We want it to be deleted on the main thread. If we were to move it do the decoder thread, it would be deleted on that thread which is no good.

Reply:
Isn't it the case that, with the current patch, we will unref audioData in the audioDecoder WorkQueue, as well as potentially refing/unrefing it in main thread in parallel?
Shouldn't we instead ASSERT that `decodeAsync()` is called in main thread and use callOnMainThread within decodeAysnc() to move back `audioData` to main thread?